lib/app/dispatcher.rb

```diff
@@ -70,6 +71,17 @@ def scheduler
                        end
       end
 
+      def job_cleanup_timer
```

Comment:
I'm not sure if we want more threads, this is +2 threads effectively to the process, one of which does read/write operations on Elasticsearch, potentially producing more concurrency problems. Can we also do this in main thread, since these operations are lightweight?

Reply:
I don't quite understand. This method starts a `TimerTask`, which will run one more thread, not two.

The main thread loops through `.elastic-connectors`, and the job cleanup is targeting `.elastic-connectors-sync-jobs`.

Reply:
TimerTask as far as I remember is actually 2 threads: 1 monitoring thread that checks if the function inside the timer should be triggered + 1 executor thread that actually runs the callback specified in the timer.

> The main thread loops through .elastic-connectors, and the job cleanup is targeting .elastic-connectors-sync-jobs.

Indeed, my comment was mostly from perspective of reduction of complexity of the system - now we have operations happening in multiple threads while in reality we can have only 1 main thread for producer that does all the logic about connectors/jobs faciliation + 1 thread per sync process.

Reply:
I re-read the [TimerTask doc](https://ruby-concurrency.github.io/concurrent-ruby/1.1.4/Concurrent/TimerTask.html), and you are right:
>When a `TimerTask` is launched it starts a thread for monitoring the execution interval. The `TimerTask` thread does not perform the task, however. Instead, the `TimerTask` launches the task on a separate thread.

However, I'm not so concerned about the race condition.
1. Orphaned jobs will be simply deleted.
2. Stuck Jobs will be updated. With proper `stuck_threshold` configured, most of, if not all, the stuck jobs are really in stuck, while the job consumer only process pending jobs (jobs in `pending` or `suspended` status).

Reply:
In the future, the cleanup task can be added to job consumer. My concern is job consumer is required to be run much more frequently than cleanup, unless we have a place to record when the last clean up runs.